### PR TITLE
Refactor importers

### PR DESCRIPTION
**Pull request by njakobsen**
Any number of importers can be specified on a per model basis. They can pull data from any number of sources.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -48,3 +48,33 @@ execute("
...
+Person.new(:features => [Feature.new(:geog => 'some binary PostGIS Geography string')])
+```
+
+### Import
+
+You can specify multiple import sources for geometry. Each key is a method that returns the data for the Importer, and
+each value is the Importer to use to parse the data. See each Importer for more details.
+```ruby
+class Location < ActiveRecord::Base
+  has_spatial_features :import => { :remote_kml_url => :KMLFile, :file => :File }
```
Comment (rywall) [inline README.md]: Convert the import symbol to snakecase.

File: app/models/feature.rb [hunk 2]
```diff
@@ -98,8 +102,21 @@ def kml(options = {})
...
   end
 
+  def make_valid?
+    @make_valid
+  end
+
   private
 
+  def make_valid
+    self.geog = ActiveRecord::Base.connection.select_value("SELECT ST_CollectionExtract('#{sanitize}',3)")
```
Comment (rywall) [inline app/models/feature.rb]: Why doesn't this call `ST_MakeValid`?

File: lib/spatial_features/has_spatial_features/feature_import.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        imports = spatial_feature_imports(import_options)
+        cache_key = Digest::MD5.hexdigest(imports.collect(&:cache_key).join)
+        if has_spatial_features_hash? && cache_key != features_hash
+          import_features(imports)
+          validate_features!(skip_invalid)
+          update_attributes(:features_hash => cache_key)
+        elsif !has_spatial_features_hash?
+          import_features(imports)
+          validate_features!(skip_invalid)
+        end
```
Comment (rywall) [inline lib/spatial_features/has_spatial_features/feature_import.rb]: Maybe we could DRY this up a bit?

```ruby
return unless updated_features?

import_features(imports)
validate_features!(skip_invalid)
update_attributes(:features_hash => cache_key) if has_spatial_features_hash?
```

File: lib/spatial_features/importers/kml_file.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        super unzip(fetch(path_or_url)), *args
+      end
+
+      private
+
+      def fetch(path_or_url)
+        normalize_file(open(path_or_url))
+      rescue SocketError, Errno::ECONNREFUSED
+        url = URI(path_or_url)
+        raise UpdateError, "ArcGIS Server is not responding. Ensure ArcGIS Server is running and accessible at #{[url.scheme, "//#{url.host}", url.port].select(&:present?).join(':')}."
```
Comment (njakobsen) [inline lib/spatial_features/importers/kml_file.rb]: Move ArcGIS-specific error messages to the ArcGIS kml importer.

File: lib/spatial_features/importers/shapefile.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        end
+      end
+
+      def geom_from_text(wkt)
+        ActiveRecord::Base.connection.select_value("SELECT ST_GeomFromText('#{wkt}')")
+      end
+
+      def unzip(file)
+        path = ::File.path(file)
+        path = Unzip.paths(file, :find => '.shp') || raise(UpdateError, "File missing SHP") if path.end_with?('.zip')
```
Comment (njakobsen) [inline lib/spatial_features/importers/shapefile.rb]: Use the `Unzip.is_zip?` method

File: lib/spatial_features/has_spatial_features/feature_import.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    def spatial_feature_imports(options = {})
+      spatial_features_options.fetch(:import, {}).collect do |data_method, importer_name|
+        data = send(data_method)
+        "SpatialFeatures::Importers::#{importer_name}".constantize.new(data, options) if data.present?
+      end.compact
+    end
+
+    def import_features(imports)
+      features.destroy_all
+      features << imports.flat_map(&:features)
```
Comment (rywall) [inline lib/spatial_features/has_spatial_features/feature_import.rb]: Why this and not

`self.features = imports.flat_map(&:features)`

File: lib/spatial_features/has_spatial_features/feature_import.rb [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    end
+
+    def validate_features!(skip_invalid = false)
+      invalid = features.select {|feature| feature.errors.present? }
+      features.destroy(invalid)
```
Comment (njakobsen) [inline lib/spatial_features/has_spatial_features/feature_import.rb]: Add test for destroying invalid features. Does it explode when it sees a new_record?

File: lib/spatial_features/importers/base.rb [hunk 8]
```diff
@@ -0,0 +1,48 @@
+require 'digest/md5'
+
+module SpatialFeatures
+  module Importers
+    class Base
+      attr_reader :errors
+
+      def initialize(data, skip_invalid: false, make_valid: false)
+        @make_valid = make_valid
+        @feature_error_messages = []
```
Comment (rywall) [inline lib/spatial_features/importers/base.rb]: What is this used for?

File: lib/spatial_features/has_spatial_features/feature_import.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
+require 'digest/md5'
+
+module SpatialFeatures
+  module FeatureImport
+    def update_features!(skip_invalid: false, **import_options)
```
Comment (njakobsen) [inline lib/spatial_features/has_spatial_features/feature_import.rb]: Make sure this method returns a truthy value if features were updated, and nil if they were not. It is used in both a save callback and in a boolean condition in connect engine.

File: lib/spatial_features/importers/base.rb [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        @data = data
+        @errors = []
```
Comment (rywall) [inline lib/spatial_features/importers/base.rb]: These should be accessed by `update_features!`

File: lib/spatial_features/importers/base.rb [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+          rescue => e
+            @errors << e.message
+          end
+        end
+
+        return new_features
+      end
+
+      def each_record(&block)
+        @data.each(&block)
```
Comment (rywall) [inline lib/spatial_features/importers/base.rb]: Maybe `raise NotImplementedError`

File: lib/spatial_features/importers/file.rb [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+      def initialize(data, *args)
+        names = Unzip.names(data)
+        if names.any? {|name| name.ends_with? '.kml' }
+          __setobj__(KMLFile.new(data, *args))
+
+        elsif names.any? {|name| name.ends_with? '.shp' }
+          __setobj__(Shapefile.new(data, *args))
+
+        else
+          raise "Could not detect importer for file"
```
Comment (rywall) [inline lib/spatial_features/importers/file.rb]: We probably want to handle this error somewhere and add a validation error to the model.

File: lib/spatial_features/importers/kml_file.rb [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+      rescue OpenURI::HTTPError
+        raise UpdateError, "ArcGIS Map Service not found. Ensure ArcGIS Server is running and accessible at #{path_or_url}."
```
Comment (rywall) [inline lib/spatial_features/importers/kml_file.rb]: Any way to easily move these to `KMLFileArcGIS` since they are ArcGIS specific?

**Pull request merged**